### ISSUE TITLE
Split up IArtifactSource

### ISSUE THREAD

**Issue by ewoutkramer** [ewoutkramer/fhir-net-api#127]
IArtifactDirectory with FindConformanceResources()
IConformanceSource with FindCRByCanonicalId()
IConformanceFilter with FindValueSetBySystem() - based on search possibilities of resource in fhir spec?

The ArtifactResolver can then determine which is the most efficient way to call the source:
- If it just implements IConformanceSource, call FindConformanceResources().Where(cr=>cr.System=xxxxx)
- If it implements IConformanceFilter call FindValueSetBySystem().

Problem now is that even if the source can quickly index and search on some aspect, you still have to list all conformance resources and then do a Where() on it. On some ArtifactSources, this has a large performance hit.

**Comment by ewoutkramer** [comment]
Maybe IConformanceFilter should have one function: Where

This takes a predicate ConformanceInformation => bool, the source itself can then decide whether to run this internally on an IQueryable (database) or IEnumerable (most other sources with in-mem caches).

**Comment by ewoutkramer** [comment]
![image](https://cloud.githubusercontent.com/assets/2276305/11095136/d0fa95b4-8893-11e5-89a8-447b13034a30.png)

**Comment by ewoutkramer** [comment]
After a new discussion we have decided that IConformanceFilter is probably too complex for most to implement.

We decided to have a set of interfaces, one for each major usecase (lookup resource by canonical id, lookup valueset by system, list resources by type), return basic metadata (as is the current situation, but maybe have an explicit IBasicConformanceInfo that's also implemented by the "real" conformance resources).  When  more (meta)data is needed, one could request the whole resource. There may be a lower-level interface that just exposes a list of IBasicConformanceInfo.

ArtifactSource will "scale up": if a source does not implement the required specific interface, it may fall back to the simpler interface. This policy may need to be configurable, since a client may want to avoid falling back to a very inefficient query, so ask the ArtifactSource to either throw or return null instead.
